File: app/lib/signal/ppg.dart
import 'dart:math' as math;
import 'dart:typed_data';

import 'filters.dart';
import 'rr_features.dart';

/// Result of analysing one contact-PPG capture (MAX30102, IR channel).
class PpgResult {
  /// Sample indices of detected systolic peaks.
  final List<int> peaks;

  /// Inter-beat intervals in milliseconds.
  final Float64List ibiMs;

  /// Pulse rate from the PPG, beats per minute.
  final double meanPulseRate;

  /// Irregularity of the pulse interval series, on the same 0-1 scale as the
  /// ECG's RR irregularity so the two are directly comparable.
  final double irregularityScore;

  /// AC/DC x 100. Below [PpgAnalyser.kPerfusionGate] the trace is untrustworthy.
  final double perfusionIndex;

  /// Normalised spread of perfusion index across 1 s sub-windows of the
  /// capture: (max - min) / mean. Zero for a steady contact, high when the
  /// pulsatile signal swings up and down within one capture — which is what
  /// happens when a finger shifts on the sensor mid-recording.
  ///
  /// Distinct from [perfusionIndex] itself: that catches a signal that is
  /// UNIFORMLY weak (cold finger, low LED current); this catches one that is
  /// unstable, which a uniformly weak signal is not. This is the PPG's
  /// contribution to inferred motion detection, now that the MPU-6050 that
  /// used to sense motion directly is no longer in the BOM. See
  /// contracts/ppg.md.
  final double perfusionStabilityRatio;

  /// Whether this capture may be used for anything at all.
  final bool usable;

  final String? failureReason;

  /// Full interval statistics, for the record and for debugging.
  final RrFeatures features;

  const PpgResult({
    required this.peaks,
    required this.ibiMs,
    required this.meanPulseRate,
    required this.irregularityScore,
    required this.perfusionIndex,
    this.perfusionStabilityRatio = 0.0,
    required this.usable,
    required this.features,
    this.failureReason,
  });

  static final PpgResult unusable = PpgResult(
    peaks: const [],
    ibiMs: Float64List(0),
    meanPulseRate: 0,
    irregularityScore: 0,
    perfusionIndex: 0,
    usable: false,
    features: RrFeatures.empty,
    failureReason: 'No usable pulse signal',
  );

  /// Coarse outcome for the pre-screen step and for the screening record.
  /// Matches the `ppgResult` enum in contracts/record.schema.json.
  String get prescreenOutcome {
    if (!usable) return 'unclear';
    if (features.count < PpgAnalyser.kMinBeats) return 'unclear';
    return irregularityScore >= PpgAnalyser.kIrregularityGate
        ? 'irregular'
        : 'regular';
  }

  Map<String, dynamic> toJson() => {
        'meanPulseRate': meanPulseRate,
        'irregularityScore': irregularityScore,
        'perfusionIndex': perfusionIndex,
        'perfusionStabilityRatio': perfusionStabilityRatio,
        'beatCount': features.count,
        'usable': usable,
        'prescreenOutcome': prescreenOutcome,
        'failureReason': failureReason,
      };
}

/// Contact PPG analysis for the MAX30102 IR channel.
///
/// Deliberately mirrors the ECG path — same filter machinery, same interval
/// statistics via [RrAnalyser] — so that pulse irregularity and RR irregularity
/// live on one scale and can be compared without a conversion factor anyone has
/// to remember.
///
/// See contracts/ppg.md.
class PpgAnalyser {
  final double fs;

  /// Perfusion index (%) below which the PPG is discarded. A cold finger, weak
  /// contact, or insufficient LED current all land here. Discarding is correct:
  /// the ECG decision stands on its own, and a bad PPG must never weaken it.
  static const double kPerfusionGate = 0.3;

  /// Minimum systolic peaks before the interval statistics mean anything.
  /// Lower than the ECG's 30 because the PPG pre-screen runs on a shorter
  /// capture — its job is to decide whether to spend an electrode, not to
  /// produce a tier.
  static const int kMinBeats = 12;

  /// Pulse rates outside this range are rejected as detection failures rather
  /// than reported as findings.
  static const double kMinIbiMs = 250; // 240 bpm
  static const double kMaxIbiMs = 2000; // 30 bpm

  /// Peaks within this distance of either end of the capture are discarded.
  ///
  /// Zero-phase filtering leaves small transients at both edges — the padding
  /// suppresses most of it, but a capture that begins or ends mid-pulse still
  /// produces a bump large enough to clear the threshold. Those are not beats.
  ///
  /// This is not cosmetic. A spurious pulse inflates the PPG rate, which drives
  /// pulseDeficitBpm NEGATIVE, which the fusion correctly reports as impossible
  /// and turns into a RETAKE — so an edge artefact would present to the worker
  /// as "beat detection unreliable" on a perfectly good capture. Losing at most
  /// one real beat at each end of a 30 s window costs nothing by comparison.
  static const double kEdgeGuardMs = 500;

  /// Same gate value as the ECG rules, on the same scale, by construction.
  static const double kIrregularityGate = 0.5;

  /// Sub-window length for [_perfusionStability]. Short enough that a finger
  /// shift shows up as a swing between windows, long enough to contain
  /// several pulses per window so a single-window perfusion index is not
  /// itself noise.
  static const double kMotionSubWindowSec = 1.0;

  static const RrAnalyser _intervals = RrAnalyser();

  PpgAnalyser({this.fs = 100});

  PpgResult analyse(Float64List rawIr) {
    if (rawIr.length < fs * 5) {
      return PpgResult.unusable;
    }

    // Perfusion index first: it is a ratio of pulsatile amplitude to the DC
    // level, so it must be measured before the highpass removes the DC.
    final perfusion = _perfusionIndex(rawIr);
    if (perfusion < kPerfusionGate) {
      return PpgResult(
        peaks: const [],
        ibiMs: Float64List(0),
        meanPulseRate: 0,
        irregularityScore: 0,
        perfusionIndex: perfusion,
        usable: false,
        features: RrFeatures.empty,
        failureReason:
            'Weak pulse signal - warm the finger, rest it gently on the sensor',
      );
    }

    final band = FilterChain.ppgBand(fs).filtfilt(rawIr);
    final peaks = _detectSystolicPeaks(band);

    if (peaks.length < 2) {
      return PpgResult(
        peaks: peaks,
        ibiMs: Float64List(0),
        meanPulseRate: 0,
        irregularityScore: 0,
        perfusionIndex: perfusion,
        usable: false,
        features: RrFeatures.empty,
        failureReason: 'Could not find a pulse - reposition the finger',
      );
    }

    final ibi = Float64List(peaks.length - 1);
    for (var i = 0; i < ibi.length; i++) {
      ibi[i] = (peaks[i + 1] - peaks[i]) * 1000.0 / fs;
    }

    // Reuse the ECG interval statistics verbatim. AF is irregular in the pulse
    // for exactly the reason it is irregular in the RR series, so the same
    // RMSSD / pNN50 / entropy combination applies unchanged.
    final feats = _intervals.analyse(ibi);

    return PpgResult(
      peaks: peaks,
      ibiMs: ibi,
      meanPulseRate: feats.meanHr,
      irregularityScore: feats.irregularityScore,
      perfusionIndex: perfusion,
      perfusionStabilityRatio: _perfusionStability(rawIr),
      usable: feats.count >= 2,
      features: feats,
    );
  }

  /// AC/DC x 100, the standard perfusion index.
  ///
  /// AC is taken as the 5th-to-95th percentile span of the pulsatile component
  /// rather than the raw peak-to-peak, so a single motion spike cannot
  /// manufacture a healthy-looking number.
  double _perfusionIndex(Float64List raw) {
    var dc = 0.0;
    for (final v in raw) {
      dc += v;
    }
    dc /= raw.length;
    if (dc.abs() < 1e-9) return 0.0;

    final ac = FilterChain.ppgBand(fs).filtfilt(raw);
    final sorted = Float64List.fromList(ac)..sort();
    final lo = sorted[(sorted.length * 0.05).floor()];
    final hi = sorted[(sorted.length * 0.95).floor()];

    return ((hi - lo).abs() / dc.abs()) * 100.0;
  }

  /// Spread of local AC amplitude across 1 s sub-windows, normalised by the
  /// capture's overall DC level.
  ///
  /// Zero for a uniformly steady contact, whether that contact is good or bad
  /// — a uniformly weak signal is [kPerfusionGate]'s job, not this one's. High
  /// only when the signal quality itself swings during the capture, which is
  /// the PPG-side signature of the finger moving.
  ///
  /// Filters the WHOLE capture once, then measures local spread on the
  /// already-filtered signal — NOT by recomputing [_perfusionIndex]
  /// independently on each short sub-window. A 0.5 Hz highpass needs on the
  /// order of seconds to settle; re-running it from scratch on an isolated
  /// 100-sample (1 s) slice produces filter-edge noise of the same order as
  /// the physiological signal being measured, which swamped this metric with
  /// spurious instability on a perfectly steady synthetic capture during
  /// validation (`ml/reference/validate_ppg.py`). Filtering once and then
  /// windowing the result is what makes the measurement mean what it says.
  double _perfusionStability(Float64List raw) {
    var dc = 0.0;
    for (final v in raw) {
      dc += v;
    }
    dc /= raw.length;
    if (dc.abs() < 1e-9) return 0.0;

    final ac = FilterChain.ppgBand(fs).filtfilt(raw);
    final window = (kMotionSubWindowSec * fs).round();
    final n = raw.length ~/ window;
    if (n < 2) return 0.0;

    final spreads = Float64List(n);
    for (var i = 0; i < n; i++) {
      // Copy before sorting, matching _perfusionIndex above: a sublistView
      // shares the underlying buffer, so sorting it in place would mutate ac
      // itself. Harmless today (disjoint ranges, ac unused after this loop),
      // but a copy is what keeps that true if this function is ever refactored.
      final seg = Float64List.fromList(
          ac.sublist(i * window, (i + 1) * window))
        ..sort();
      final lo = seg[(seg.length * 0.05).floor()];
      final hi = seg[(seg.length * 0.95).floor()];
      spreads[i] = (hi - lo).abs() / dc.abs() * 100.0;
    }

    var mean = 0.0;
    for (final s in spreads) {
      mean += s;
    }
    mean /= spreads.length;
    if (mean < 1e-9) return 0.0;

    var maxV = spreads.first, minV = spreads.first;
    for (final s in spreads) {
      if (s > maxV) maxV = s;
      if (s < minV) minV = s;
    }
    return (maxV - minV) / mean;
  }

  /// Systolic peak detection on the bandpassed signal.
  ///
  /// A PPG pulse is a broad, smooth hump rather than the sharp spike of a QRS,
  /// so Pan-Tompkins' derivative-and-square emphasis is the wrong tool. An
  /// adaptive amplitude threshold with a physiological refractory period suits
  /// the waveform shape better and has far fewer ways to go wrong.
  List<int> _detectSystolicPeaks(Float64List x) {
    final refractory = math.max(1, (kMinIbiMs / 1000.0 * fs).round());

    // Threshold from the signal's own distribution: robust to LED current and
    // to how hard the patient is pressing, neither of which we control.
    final sorted = Float64List.fromList(x)..sort();
    final median = sorted[sorted.length ~/ 2];
    final upper = sorted[(sorted.length * 0.75).floor()];
    final threshold = median + 0.5 * (upper - median);

    final guard = (kEdgeGuardMs / 1000.0 * fs).round();
    final peaks = <int>[];
    for (var i = 1; i < x.length - 1; i++) {
      if (i < guard || i >= x.length - guard) continue;
      if (x[i] <= threshold) continue;
      if (x[i] <= x[i - 1] || x[i] < x[i + 1]) continue;
      if (peaks.isNotEmpty && i - peaks.last < refractory) {
        if (x[i] > x[peaks.last]) peaks[peaks.length - 1] = i;
        continue;
      }
      peaks.add(i);
    }
    return peaks;
  }
}

File: android/app/src/main/kotlin/com/arogyax/signal/Ppg.kt
package com.arogyax.signal

import kotlin.math.max
import kotlin.math.min

/**
 * Result of analysing one contact-PPG capture (MAX30102, IR channel).
 *
 * Port of app/lib/signal/ppg.dart's PpgResult - keep the two in sync.
 */
data class PpgResult(
    /** Sample indices of detected systolic peaks. */
    val peaks: List<Int>,
    /** Inter-beat intervals in milliseconds. */
    val ibiMs: DoubleArray,
    /** Pulse rate from the PPG, beats per minute. */
    val meanPulseRate: Double,
    /** Irregularity of the pulse interval series, on the same 0-1 scale as the ECG's RR irregularity. */
    val irregularityScore: Double,
    /** AC/DC x 100. Below [PpgAnalyser.K_PERFUSION_GATE] the trace is untrustworthy. */
    val perfusionIndex: Double,
    /**
     * Normalised spread of perfusion index across 1 s sub-windows of the
     * capture: (max - min) / mean. Zero for a steady contact, high when the
     * pulsatile signal swings up and down within one capture - the PPG's
     * contribution to inferred motion detection, now that the MPU-6050 is no
     * longer in the BOM. See contracts/ppg.md.
     */
    val perfusionStabilityRatio: Double = 0.0,
    /** Whether this capture may be used for anything at all. */
    val usable: Boolean,
    /** Full interval statistics, for the record and for debugging. */
    val features: RrFeatures,
    val failureReason: String? = null,
) {
    /**
     * Coarse outcome for the pre-screen step and for the screening record.
     * Matches the `ppgResult` enum in contracts/record.schema.json.
     */
    val prescreenOutcome: String
        get() {
            if (!usable) return "unclear"
            if (features.count < PpgAnalyser.K_MIN_BEATS) return "unclear"
            return if (irregularityScore >= PpgAnalyser.K_IRREGULARITY_GATE) "irregular" else "regular"
        }

    companion object {
        val UNUSABLE = PpgResult(
            peaks = emptyList(),
            ibiMs = DoubleArray(0),
            meanPulseRate = 0.0,
            irregularityScore = 0.0,
            perfusionIndex = 0.0,
            usable = false,
            features = RrFeatures.EMPTY,
            failureReason = "No usable pulse signal",
        )
    }
}

/**
 * Contact PPG analysis for the MAX30102 IR channel.
 *
 * Deliberately mirrors the ECG path - same filter machinery, same interval
 * statistics via [RrAnalyser] - so that pulse irregularity and RR irregularity
 * live on one scale and can be compared without a conversion factor anyone
 * has to remember. See contracts/ppg.md.
 *
 * Port of app/lib/signal/ppg.dart's PpgAnalyser - keep the two in sync.
 */
class PpgAnalyser(private val fs: Double = 100.0) {
    private val intervals = RrAnalyser()

    fun analyse(rawIr: DoubleArray): PpgResult {
        if (rawIr.size < fs * 5) return PpgResult.UNUSABLE

        // Perfusion index first: it is a ratio of pulsatile amplitude to the
        // DC level, so it must be measured before the highpass removes the DC.
        val perfusion = perfusionIndex(rawIr)
        if (perfusion < K_PERFUSION_GATE) {
            return PpgResult(
                peaks = emptyList(),
                ibiMs = DoubleArray(0),
                meanPulseRate = 0.0,
                irregularityScore = 0.0,
                perfusionIndex = perfusion,
                usable = false,
                features = RrFeatures.EMPTY,
                failureReason = "Weak pulse signal - warm the finger, rest it gently on the sensor",
            )
        }

        val band = FilterChain.ppgBand(fs).filtfilt(rawIr)
        val peaks = detectSystolicPeaks(band)

        if (peaks.size < 2) {
            return PpgResult(
                peaks = peaks,
                ibiMs = DoubleArray(0),
                meanPulseRate = 0.0,
                irregularityScore = 0.0,
                perfusionIndex = perfusion,
                usable = false,
                features = RrFeatures.EMPTY,
                failureReason = "Could not find a pulse - reposition the finger",
            )
        }

        val ibi = DoubleArray(peaks.size - 1) { (peaks[it + 1] - peaks[it]) * 1000.0 / fs }

        // Reuse the ECG interval statistics verbatim. AF is irregular in the
        // pulse for exactly the reason it is irregular in the RR series, so
        // the same RMSSD / pNN50 / entropy combination applies unchanged.
        val feats = intervals.analyse(ibi)

        return PpgResult(
            peaks = peaks,
            ibiMs = ibi,
            meanPulseRate = feats.meanHr,
            irregularityScore = feats.irregularityScore,
            perfusionIndex = perfusion,
            perfusionStabilityRatio = perfusionStability(rawIr),
            usable = feats.count >= 2,
            features = feats,
        )
    }

    /**
     * AC/DC x 100, the standard perfusion index.
     *
     * AC is taken as the 5th-to-95th percentile span of the pulsatile
     * component rather than the raw peak-to-peak, so a single motion spike
     * cannot manufacture a healthy-looking number.
     */
    private fun perfusionIndex(raw: DoubleArray): Double {
        var dc = 0.0
        for (v in raw) dc += v
        dc /= raw.size
        if (kotlin.math.abs(dc) < 1e-9) return 0.0

        val ac = FilterChain.ppgBand(fs).filtfilt(raw)
        val sorted = ac.copyOf().also { it.sort() }
        val lo = sorted[(sorted.size * 0.05).toInt()]
        val hi = sorted[(sorted.size * 0.95).toInt()]

        return kotlin.math.abs(hi - lo) / kotlin.math.abs(dc) * 100.0
    }

    /**
     * Spread of local AC amplitude across 1 s sub-windows, normalised by the
     * capture's overall DC level. See the Dart doc comment on this method in
     * ppg.dart for the full explanation of why the whole capture is filtered
     * once rather than re-filtering each short sub-window independently.
     */
    private fun perfusionStability(raw: DoubleArray): Double {
        var dc = 0.0
        for (v in raw) dc += v
        dc /= raw.size
        if (kotlin.math.abs(dc) < 1e-9) return 0.0

        val ac = FilterChain.ppgBand(fs).filtfilt(raw)
        val window = Math.round(K_MOTION_SUB_WINDOW_SEC * fs).toInt()
        val n = raw.size / window
        if (n < 2) return 0.0

        val spreads = DoubleArray(n)
        for (i in 0 until n) {
            // Copy before sorting, matching perfusionIndex above.
            val seg = ac.copyOfRange(i * window, (i + 1) * window).also { it.sort() }
            val lo = seg[(seg.size * 0.05).toInt()]
            val hi = seg[(seg.size * 0.95).toInt()]
            spreads[i] = kotlin.math.abs(hi - lo) / kotlin.math.abs(dc) * 100.0
        }

        var mean = 0.0
        for (s in spreads) mean += s
        mean /= spreads.size
        if (mean < 1e-9) return 0.0

        var maxV = spreads[0]
        var minV = spreads[0]
        for (s in spreads) {
            if (s > maxV) maxV = s
            if (s < minV) minV = s
        }
        return (maxV - minV) / mean
    }

    /**
     * Systolic peak detection on the bandpassed signal.
     *
     * A PPG pulse is a broad, smooth hump rather than the sharp spike of a
     * QRS, so Pan-Tompkins' derivative-and-square emphasis is the wrong tool.
     * An adaptive amplitude threshold with a physiological refractory period
     * suits the waveform shape better and has far fewer ways to go wrong.
     */
    private fun detectSystolicPeaks(x: DoubleArray): List<Int> {
        val refractory = max(1, Math.round(K_MIN_IBI_MS / 1000.0 * fs).toInt())

        // Threshold from the signal's own distribution: robust to LED current
        // and to how hard the patient is pressing, neither of which we control.
        val sorted = x.copyOf().also { it.sort() }
        val median = sorted[sorted.size / 2]
        val upper = sorted[(sorted.size * 0.75).toInt()]
        val threshold = median + 0.5 * (upper - median)

        val guard = Math.round(K_EDGE_GUARD_MS / 1000.0 * fs).toInt()
        val peaks = mutableListOf<Int>()
        for (i in 1 until x.size - 1) {
            if (i < guard || i >= x.size - guard) continue
            if (x[i] <= threshold) continue
            if (x[i] <= x[i - 1] || x[i] < x[i + 1]) continue
            if (peaks.isNotEmpty() && i - peaks.last() < refractory) {
                if (x[i] > x[peaks.last()]) peaks[peaks.size - 1] = i
                continue
            }
            peaks.add(i)
        }
        return peaks
    }

    companion object {
        /**
         * Perfusion index (%) below which the PPG is discarded. A cold
         * finger, weak contact, or insufficient LED current all land here.
         * Discarding is correct: the ECG decision stands on its own, and a
         * bad PPG must never weaken it.
         */
        const val K_PERFUSION_GATE = 0.3

        /**
         * Minimum systolic peaks before the interval statistics mean
         * anything. Lower than the ECG's 30 because the PPG pre-screen runs
         * on a shorter capture.
         */
        const val K_MIN_BEATS = 12

        /** Pulse rates outside this range are rejected as detection failures rather than findings. */
        const val K_MIN_IBI_MS = 250.0 // 240 bpm
        const val K_MAX_IBI_MS = 2000.0 // 30 bpm

        /**
         * Peaks within this distance of either end of the capture are
         * discarded - zero-phase filtering leaves edge transients large
         * enough to look like beats. See ppg.dart's doc comment on why this
         * matters (it is not cosmetic - it prevents a false "beat detection
         * unreliable" retake on an otherwise good capture).
         */
        const val K_EDGE_GUARD_MS = 500.0

        /** Same gate value as the ECG rules, on the same scale, by construction. */
        const val K_IRREGULARITY_GATE = 0.5

        /** Sub-window length for perfusion stability. */
        const val K_MOTION_SUB_WINDOW_SEC = 1.0
    }
}
